driver-core/src/main/com/mongodb/client/gridfs/model/GridFSFile.java

```diff
@@ -0,0 +1,253 @@
+import java.util.List;
+
+import static com.mongodb.assertions.Assertions.notNull;
+
+/**
+ * The GridFSFile
+ *
+ * @since 3.1
+ */
+public class GridFSFile {
+    private final BsonValue id;
+    private final String filename;
+    private final long length;
+    private final int chunkSize;
+    private final Date uploadDate;
+    private final String md5;
+
+    // Optional values
+    private final Document metadata;
+
+    // Deprecated values
+    private final String contentType;
+    private final List<String> aliases;
+
+    /**
+     * Creates a new GridFSFile
+     *
+     * @param id the id of the file
+     * @param filename the filename
+     * @param length the length, in bytes of the file
+     * @param chunkSize the chunkSize, in bytes of the file
+     * @param uploadDate the upload date of the file
+     * @param md5 the hash of the files contents
+     * @param metadata the optional metadata for the file
+     */
+    public GridFSFile(final BsonValue id, final String filename, final long length, final int chunkSize, final Date uploadDate,
+                      final String md5, final Document metadata) {
+        this.id = notNull("id", id);
+        this.filename = notNull("filename", filename);
+        this.length = notNull("length", length);
+        this.chunkSize = notNull("chunkSize", chunkSize);
+        this.uploadDate = notNull("uploadDate", uploadDate);
+        this.md5 = notNull("md5", md5);
+        this.metadata = metadata;
+        this.contentType = null;
+        this.aliases = null;
+    }
+
+    /**
+     * Creates a legacy implementation of the GridFSFile
+     *
+     * <p>For GridFS files created in older versions of the driver.</p>
+     *
+     * @param id the id of the file
+     * @param filename the filename
+     * @param length the length, in bytes of the file
+     * @param chunkSize the chunkSize, in bytes of the file
+     * @param uploadDate the upload date of the file
+     * @param md5 the hash of the files contents
+     * @param metadata the optional metadata for the file
+     * @param contentType the optional contentType for the file
+     * @param aliases the optional aliases for the file
+     */
+    public GridFSFile(final BsonValue id, final String filename, final long length, final int chunkSize, final Date uploadDate,
+                      final String md5, final Document metadata, final String contentType, final List<String> aliases) {
+        this.id = notNull("id", id);
+        this.filename = notNull("filename", filename);
+        this.length = notNull("length", length);
+        this.chunkSize = notNull("chunkSize", chunkSize);
+        this.uploadDate = notNull("uploadDate", uploadDate);
+        this.md5 = notNull("md5", md5);
+        this.metadata = metadata;
+        this.contentType = contentType;
+        this.aliases = aliases;
+    }
+
+    /**
+     * The id for this file.
+     *
+     * @return the id for this file
+     */
+    public BsonValue getId() {
```

Comment:
Maybe add:

    public ObjectId getObjectId()

and deprecate getId(), since using a non-ObjectId is deprecated.